### PR TITLE
Envoy Dynamic SDS ConfigDump.

### PR DESCRIPTION
**Pull request by incfly**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Support Dynamic SDS Config Dump
Risk Level: Medium?
Testing: Unit test
Docs Changes: Not sure if config dump requires documentation
Release Notes: config_dump adds secrets fetched from SDS
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by incfly** [comment]
@JimmyCYJ

File: source/common/secret/secret_manager_impl.h [hunk 1]
```diff
@@ -68,6 +71,17 @@ class SecretManagerImpl : public SecretManager {
       return secret_provider;
     }
 
+    std::vector<std::shared_ptr<SecretType>> allSecrets() {
+      std::vector<std::shared_ptr<SecretType>> secrets;
+      for (const auto& secret_entry : dynamic_secret_providers_) {
+        std::shared_ptr<SecretType> secret_provider = secret_entry.second.lock();
+        if (secret_provider) {
+          secrets.push_back(secret_provider);
```
Comment (silentdai) [inline source/common/secret/secret_manager_impl.h]: avoid atomic_inc and atomic_dec inside shared_ptr
secret.push_back(std::move(secret_provider));

File: source/common/secret/sds_api.h [hunk 2]
```diff
@@ -68,6 +78,8 @@ class SdsApi : public Config::SubscriptionCallbacks {
   Cleanup clean_up_;
   ProtobufMessage::ValidationVisitor& validation_visitor_;
   Config::SubscriptionFactory& subscription_factory_;
+  TimeSource& time_source_;
+  SecretData secret_data_;
```
Comment (silentdai) [inline source/common/secret/sds_api.h]: I am wondering if we need it as a member.
It seems it's internal fields are populated by `dumpSecretConfigs`
Could you explain why we cannot use a fresh `SecretData` on stack of dumpSecretConfigs?

File: source/common/secret/sds_api.h [hunk 3]
```diff
@@ -32,11 +32,21 @@ namespace Secret {
...
+    SystemTime last_updated_;
+  };
+
   SdsApi(envoy::api::v2::core::ConfigSource sds_config, absl::string_view sds_config_name,
-         Config::SubscriptionFactory& subscription_factory,
+         Config::SubscriptionFactory& subscription_factory, TimeSource& time_source,
          ProtobufMessage::ValidationVisitor& validation_visitor, Stats::Store& stats,
          Init::Manager& init_manager, std::function<void()> destructor_cb);
 
+  // TODO: what's the thread model of the admin::config_dump vs xDS config updates?
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/secret/sds_api.h]: xDS config updates happen at main thread, I didn't check /config_dump but I think it should be served by main thread as well. I will double check.

**Comment by incfly** [comment]
It appears to be on main thread,
https://github.com/envoyproxy/envoy/blob/master/source/exe/main_common.h#L52

So we are good without worry about racing, i guess?

File: source/common/secret/sds_api.h (hunk 2 above)
Comment (incfly) [reply source/common/secret/sds_api.h]: dumnpSecretConfigs is SecretManagerImpl's method, which owns reference of sdsapi object.

SdsApi is updated when subscription updates event occurs, right? if no member here, how to track the these information when config udpate occurs?

File: source/common/secret/secret_manager_impl.h (hunk 1 above)
Comment (incfly) [reply source/common/secret/secret_manager_impl.h]: changed, but curious, do you mean the std::shared_ptr copy constructor internally uses atomic_inc stuff to track the counting? failed to find any doc, will be intersted to see some reference :)

File: source/common/secret/secret_manager_impl.cc [hunk 4]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
                                                     secret_provider_context);
 }
 
+// TODO: question, what's the handling of static inlined stuff? they're just internal constructs
```
[the remark this answers is not in the document]
Comment (incfly) [reply source/common/secret/secret_manager_impl.cc]: oops, poor comment, we seems have inlined key/cert managed by secretmanagerimpl. They are not fetched by SDS dynamically per se. Not sure if we need to add them to debug page.

@JimmyCYJ thoughts?

File: source/common/secret/secret_manager_impl.cc [hunk 5]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
...
+
+  // Handling validation Context provided via SDS.
+  auto context_secrets = validation_context_providers_.allSecrets();
+  for (const auto& validation_context_secret : context_secrets) {
+    auto secret_data = validation_context_secret->secretData();
+    auto validation_context = validation_context_secret->secret();
+    auto dynamic_secret = config_dump->mutable_dynamic_secrets()->Add();
+    auto secret = dynamic_secret->mutable_secret();
+    ProtobufWkt::Timestamp last_updated_ts;
+    TimestampUtil::systemClockToTimestamp(secret_data.last_updated_, last_updated_ts);
```
[the remark this answers is not in the document]
Comment (incfly) [reply source/common/secret/secret_manager_impl.cc]: done.

File: source/common/secret/secret_manager_impl.cc [hunk 6]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
                                                     secret_provider_context);
 }
 
+// TODO: question, what's the handling of static inlined stuff? they're just internal constructs
+// not needed to be exposed, maybe?
+ProtobufTypes::MessagePtr SecretManagerImpl::dumpSecretConfigs() {
+  auto config_dump = std::make_unique<envoy::admin::v2alpha::SecretsConfigDump>();
+  auto secrets = certificate_providers_.allSecrets();
+  for (const auto& cert_secrets : secrets) {
+    auto secret_data = cert_secrets->secretData();
```
[the remark this answers is not in the document]
Comment (incfly) [reply source/common/secret/secret_manager_impl.cc]: _done._

File: source/common/secret/sds_api.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (incfly) [reply source/common/secret/sds_api.h]: Checked the main_common.cc seems invoked from main thread.

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 7]
```diff
@@ -7,6 +7,7 @@ option java_multiple_files = true;
 option java_package = "io.envoyproxy.envoy.admin.v2alpha";
 
 import "envoy/api/v2/cds.proto";
+import "envoy/api/v2/auth/cert.proto";
```
[the remark this answers is not in the document]
Comment (incfly) [reply api/envoy/admin/v2alpha/config_dump.proto]: good catch, done.

File: source/common/secret/secret_manager_impl.h [hunk 8]
```diff
@@ -68,6 +71,17 @@ class SecretManagerImpl : public SecretManager {
       return secret_provider;
     }
 
+    std::vector<std::shared_ptr<SecretType>> allSecrets() {
+      std::vector<std::shared_ptr<SecretType>> secrets;
+      for (const auto& secret_entry : dynamic_secret_providers_) {
+        std::shared_ptr<SecretType> secret_provider = secret_entry.second.lock();
```
Comment (JimmyCYJ) [inline source/common/secret/secret_manager_impl.h]: Before we call lock(), shall we check secret_entry.second.expired(), and skip expired secret_entry?

File: source/common/secret/secret_manager_impl.h [hunk 9]
```diff
@@ -68,6 +71,17 @@ class SecretManagerImpl : public SecretManager {
       return secret_provider;
     }
 
+    std::vector<std::shared_ptr<SecretType>> allSecrets() {
```
Comment (JimmyCYJ) [inline source/common/secret/secret_manager_impl.h]: rename to secretProviders()?

File: source/common/secret/secret_manager_impl.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/secret/secret_manager_impl.cc]: inline secret and static secret are not dynamic secrets. inline secrets are those file mount secrets. If multiple Listener and Cluster use the same file mount secrets, we need to copy the file mount path all over the places. Static secret simplifies this by adding a the file mount path into envoy.api.v2.auth.Secret config, and put that config into static config section of bootstrap config, other Listeners could contain a sds config that only has a sds name inside. The sds name matches the name in envoy.api.v2.auth.Secret config.

sds_static_integration_test.cc would help to explain.

File: source/common/secret/secret_manager_impl.cc [hunk 10]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
                                                     secret_provider_context);
 }
 
+// TODO: question, what's the handling of static inlined stuff? they're just internal constructs
+// not needed to be exposed, maybe?
+ProtobufTypes::MessagePtr SecretManagerImpl::dumpSecretConfigs() {
+  auto config_dump = std::make_unique<envoy::admin::v2alpha::SecretsConfigDump>();
+  auto secrets = certificate_providers_.allSecrets();
```
Comment (JimmyCYJ) [inline source/common/secret/secret_manager_impl.cc]: My concern is copying shared pointers around may not be efficient.

File: source/common/secret/secret_manager_impl.cc [hunk 11]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
...
+    const auto& dynamic_secret = config_dump->mutable_dynamic_secrets()->Add();
+    const auto& secret = dynamic_secret->mutable_secret();
+
+    ProtobufWkt::Timestamp last_updated_ts;
+    TimestampUtil::systemClockToTimestamp(secret_data.last_updated_, last_updated_ts);
+    dynamic_secret->set_version_info(secret_data.version_info_);
+    *dynamic_secret->mutable_last_updated() = last_updated_ts;
+    secret->set_name(secret_data.resource_name);
+    // TODO(incfly): handling this should be log the secret data but not the cert to signify
+    // stuck at sds updates level?
```
Comment (JimmyCYJ) [inline source/common/secret/secret_manager_impl.cc]: I don't understand this TODO. Are you asking if accessing the tls_cert during sds update would cause trouble?

File: source/common/secret/secret_manager_impl.cc [hunk 12]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
...
+    *dynamic_secret->mutable_last_updated() = last_updated_ts;
+    secret->set_name(secret_data.resource_name);
+    if (!validation_context) {
+      continue;
+    }
+    auto dump_context = secret->mutable_validation_context();
+    dump_context->MergeFrom(*validation_context);
+  }
+  return config_dump;
+}
```
Comment (JimmyCYJ) [inline source/common/secret/secret_manager_impl.cc]: Any chance to combine these two blocks into one to avoid duplicate code?

File: source/common/secret/secret_manager_impl.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/secret/secret_manager_impl.cc]: +1

File: test/common/secret/secret_manager_impl_test.cc [hunk 13]
```diff
@@ -199,6 +217,113 @@ name: "abc.com"
             tls_config.privateKey());
 }
 
+TEST_F(SecretManagerImplTest, ConfigDumpHandler) {
```
Comment (JimmyCYJ) [inline test/common/secret/secret_manager_impl_test.cc]: Please add comment to briefly introduce this test.

File: source/common/secret/secret_manager_impl.cc [hunk 14]
```diff
@@ -79,5 +84,53 @@ SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
...
+    secret->set_name(secret_data.resource_name);
+    // TODO(incfly): handling this should be log the secret data but not the cert to signify
+    // stuck at sds updates level?
+    if (!tls_cert) {
+      continue;
+    }
+    auto tls_certificate = secret->mutable_tls_certificate();
+    tls_certificate->MergeFrom(*tls_cert);
+    tls_certificate->clear_private_key();
+    tls_certificate->clear_password();
```
Comment (JimmyCYJ) [inline source/common/secret/secret_manager_impl.cc]: Maybe just copy tls_certificate->certificate_chain()?

File: test/mocks/config/mocks.cc [hunk 15]
```diff
@@ -17,6 +17,8 @@ MockSubscriptionFactory::MockSubscriptionFactory() {
         auto ret = std::make_unique<testing::NiceMock<MockSubscription>>();
         subscription_ = ret.get();
         callbacks_ = &callbacks;
+        // uint64_t hash = MessageUtil::hash(config);
+        // callbacks_map_[hash] = &callbacks;
```
Comment (JimmyCYJ) [inline test/mocks/config/mocks.cc]: Remove these?

File: test/mocks/config/mocks.h [hunk 16]
```diff
@@ -59,6 +60,9 @@ class MockSubscriptionFactory : public SubscriptionFactory {
   MOCK_METHOD0(messageValidationVisitor, ProtobufMessage::ValidationVisitor&());
 
   MockSubscription* subscription_{};
+
+  // map from the config hash to the callbacks.
+  // std::unordered_map<uint64_t, SubscriptionCallbacks*> callbacks_map_;
```
Comment (JimmyCYJ) [inline test/mocks/config/mocks.h]: Remove these?

File: test/mocks/server/mocks.h [hunk 17]
```diff
@@ -489,9 +490,10 @@ class MockTransportSocketFactoryContext : public TransportSocketFactoryContext {
   MockTransportSocketFactoryContext();
   ~MockTransportSocketFactoryContext();
 
-  Secret::SecretManager& secretManager() override { return *(secret_manager_.get()); }
+  // Secret::SecretManager& secretManager() override { return *(secret_manager_.get()); }
```
Comment (JimmyCYJ) [inline test/mocks/server/mocks.h]: Clean up?

File: test/integration/integration_admin_test.cc [hunk 18]
```diff
@@ -403,12 +404,14 @@ TEST_P(IntegrationAdminTest, Admin) {
...
   TestUtility::loadFromJson(response->body(), config_dump);
-  EXPECT_EQ(5, config_dump.configs_size());
+  EXPECT_EQ(6, config_dump.configs_size());
 
   // .. and that we can unpack one of the entries.
   envoy::admin::v2alpha::RoutesConfigDump route_config_dump;
   config_dump.configs(4).UnpackTo(&route_config_dump);
   EXPECT_EQ("route_config_0", route_config_dump.static_route_configs(0).route_config().name());
+
+  // .. TODO, do we need to add parsing and data input for the sds?
```
Comment (JimmyCYJ) [inline test/integration/integration_admin_test.cc]: Probably yes, depending on the amount of work. Could be in a separate PR.

File: test/common/secret/secret_manager_impl_test.cc [hunk 19]
```diff
@@ -199,6 +217,113 @@ name: "abc.com"
...
+    nanos: 234000000
+  secret:
+    name: "abc.com"
+    tls_certificate:
+      certificate_chain:
+        inline_string: "DUMMY_INLINE_BYTES_FOR_CERT_CHAIN"
+)EOF";
+  checkConfigDump(expected_secrets_config_dump);
+
+  // Add a dynamic tls validatoin context provider.
```
Comment (JimmyCYJ) [inline test/common/secret/secret_manager_impl_test.cc]: This test is too big! I would suggest splitting this test into two. TLS certificates and certificate validation context.

**Comment by lizan** [comment]
Can you guys just open a draft PR to upstream?

File: source/common/secret/secret_manager_impl.h (hunk 8 above)
Comment (incfly) [reply source/common/secret/secret_manager_impl.h]: https://en.cppreference.com/w/cpp/memory/weak_ptr/lock
lock should already call that under the hood.

File: source/common/secret/secret_manager_impl.h (hunk 9 above)
Comment (incfly) [reply source/common/secret/secret_manager_impl.h]: done.

File: source/common/secret/secret_manager_impl.cc (hunk 11 above)
Comment (incfly) [reply source/common/secret/secret_manager_impl.cc]: i reworded a bit.

File: test/mocks/config/mocks.h (hunk 16 above)
Comment (incfly) [reply test/mocks/config/mocks.h]: I have some questions about these... I think we need these eventually.. but unsure.

File: test/mocks/server/mocks.h (hunk 17 above)
Comment (incfly) [reply test/mocks/server/mocks.h]: done.

File: test/common/secret/secret_manager_impl_test.cc (hunk 19 above)
Comment (incfly) [reply test/common/secret/secret_manager_impl_test.cc]: i think another way can be wrapping some mock setup in a util method in the base class though

**Comment by incfly** [comment]
@lizan sure, was hesitating about the unfamiliarity of the convention by then...

https://github.com/envoyproxy/envoy/pull/7365
